docs/source/notes/extending.rst

```diff
@@ -12,53 +12,97 @@ Extending :mod:`torch.autograd`
+return either a single :class:`Tensor` output, or a :class:`tuple` of
+:class:`Tensor`s if there are multiple outputs. Also, please refer to the
+docs of :class:`Function` to find descriptions of useful methods that can be
+called only from :meth:`~Function.forward`.
+- :meth:`~Function.backward` defines the gradient formula. It will be given
+as many :class:`Tensor` arguments as there were outputs, with each of them
+representing gradient w.r.t. that output. It should return as many
+:class:`Tensor` s as there were inputs, with each of them containing the
+gradient w.r.t. its corresponding input. If your inputs didn't require
+gradient (:attr:`~ctx.needs_input_grad` is a tuple of booleans indicating
+whether each input needs gradient computation), or were non-:class:`Tensor`
+objects, you can return :class:`python:None`. Also, if you have optional
+arguments to :meth:`~Function.forward` you can return more gradients than there
+were inputs, as long as they're all :any:`python:None`.
+
+**Step 2:** It's the user's responsibility to use the functions in the forward's `ctx`
+properly in order to ensure that the new :class:`Function` works properly with
+the autograd engine.
+
+- :meth:`~torch.autograd.function.ctx.save_for_backward` must be
+  used when saving input or output of the forward to be used later in the backward.
+- :meth:`~torch.autograd.function.ctx.mark_dirty` must be used to
+  mark any input that is modified inplace by the forward function.
+- :meth:`~torch.autograd.function.ctx.mark_non_differentiable` must
+  be used to tell the engine if an output is not differentiable.
+- :meth:`~torch.autograd.function.ctx.set_materialize_grads` can be
+  used to tell the autograd engine to optimize gradient computations in the cases where
+  the output does not depend on the input by not materializing grad tensors given to backward
+  function. That is, if set to False, None object in python or "undefined tensor" (tensor x for
+  which x.defined() is False) in C++ will not be converted to a tensor filled with zeros prior
+  to calling backward. However, supporting this optimization means your custom autograd function
+  has to handle gradients that are represented in this way and is thus opt-in. Default value is True.
+
+**Step 3:** If your `~Function` does not support double backward it is
+recommended that you explicitly declare this by decorating backward with the
+:func:`~function.once_differentiable`. With this decorator, attempts to
+perform double backward through your function will produce an error.
+See our double backward tutorial for more information on double backward.
+
+**Step 4:** It is recommended that you use :func:`torch.autograd.gradcheck`
+to check whether your backward function correctly computes gradients of the
+forward by computing the Jacobian matrix using your backward function and
+comparing the value element-wise with the Jacobian computed numerically using
+finite-differencing.
 
 .. note::
```

Comment:
Is this note still necessary?

Reply:
Merging it into the section about mark_non_differentiable